Numeric check for pivot being zero in Bareiss determinant method

normalhuman opened the pull request:
#### References to other Issues or PRs

Fixes #13835

#### Brief description of what is fixed or changed

Currently, the default determinant computation method, Bareiss, accepts as a pivot any entry `!= 0`. This leads to "nan" output when the pivot is a symbolic expression that simplifies to zero. The added check to `_find_pivot` tests it for `== 0` after substituting random rational numbers for any variables. This has no noticeable impact on speed, and prevents zero pivot choices in practice. There is a small chance of rejecting a valid pivot, though.

#### Other comments

A comment next to `_find_pivot` method suggest using `_find_reasonable_pivot` instead, which implements zero detection. A note in the code says it's blocked by #12362 but I cannot reproduce #12362 and there are no errors thrown if using `_find_reasonable_pivot`. So I considered doing that but found the performance impact to be quite high. 

My timing test was a symbolic dense 7 by 7 matrix, similar to the example in #13835 but with nonzero determinant:

```
M = Matrix([[i + a*j + S(1)/(i+j+1) for i in range(7)] for j in range(7)])
```

Timing of a loop with 20 `M.det()`

 - Bareiss 165 seconds
 - Bareiss with `_find_reasonable_pivot`: 423 seconds
 - LU: 0.26 seconds (In fairness, Bareiss returns a nice formula `-391*a/1758426061161369600000 + 107/187991731629615513600000` while the output of LU is enormous and simplify hangs on it for much longer than the entire computation with the Bareiss method).

Increasing the execution time by the factor of 2.5 seemed undesirable, which is why the random numeric test was chosen instead. Timing test showed no measurable impact on performance. 

Numeric zero testing was also considered in #10279.

normalhuman committed:
Numeric check for pivot being zero in Bareiss determinant method

Fixes #13835

Currently, the default determinant computation method, Bareiss,
accepts as a pivot any entry `!= 0`. This leads to "nan" output when
the pivot is a symbolic expression that simplifies to zero.
The added check to `_find_pivot` tests it for `== 0` after
substituting random rational numbers for any variables.
This has no noticeable impact on speed, and prevents zero pivot
choices in practice.

normalhuman committed:
Test at multiple points to avoid erroneous rejection of good pivot

The number of points to try is negotiable, with five I still see only
minimal (2-3%) slowdown of .det method.

smichr commented on sympy/matrices/matrices.py:
```diff
@@ -174,10 +175,24 @@ def _eval_det_bareiss(self):
+                    if len(variables) > 0:
+                        for _ in range(5):
+                            substitutions = {v: random.randint(-10, 10) + Rational(
+                                random.randint(100, 200), random.randint(201, 300)) \
```
see also `Expr._random`. An alternative would be to give `zero` a `quick` option so it would exit before it starts any heavy lifting.

normalhuman replied:
`Expr._random()` uses floats and tries to compute the expression with precision, which can lead to serious performance issues with complex expressions where a lot of cancellation takes place. This is an example extracted from the determinant that in the linked issue:
```
var('a')
expr = ((-4*a*(-4*a*(a + 1) + a*(4*a + 1)) + a*(-4*a*(a + 4) + a*(4*a + 4)))*(-2*a*(-2*a*(a + 1) + a*(2*a + 1)) + a*(-2*a*(a + 2) + a*(2*a + 2)))/a**2 - (-2*a*(-4*a*(a + 1) + a*(4*a + 1)) + a*(-4*a*(a + 2) + a*(4*a + 2)))*(-4*a*(-2*a*(a + 1) + a*(2*a + 1)) + a*(-2*a*(a + 4) + a*(2*a + 4)))/a**2)*((-3*a*(-3*a*(a + 1) + a*(3*a + 1)) + a*(-3*a*(a + 3) + a*(3*a + 3)))*(-2*a*(-2*a*(a + 1) + a*(2*a + 1)) + a*(-2*a*(a + 2) + a*(2*a + 2)))/a**2 - (-2*a*(-3*a*(a + 1) + a*(3*a + 1)) + a*(-3*a*(a + 2) + a*(3*a + 2)))*(-3*a*(-2*a*(a + 1) + a*(2*a + 1)) + a*(-2*a*(a + 3) + a*(2*a + 3)))/a**2)/a**2 - ((-3*a*(-4*a*(a + 1) + a*(4*a + 1)) + a*(-4*a*(a + 3) + a*(4*a + 3)))*(-2*a*(-2*a*(a + 1) + a*(2*a + 1)) + a*(-2*a*(a + 2) + a*(2*a + 2)))/a**2 - (-2*a*(-4*a*(a + 1) + a*(4*a + 1)) + a*(-4*a*(a + 2) + a*(4*a + 2)))*(-3*a*(-2*a*(a + 1) + a*(2*a + 1)) + a*(-2*a*(a + 3) + a*(2*a + 3)))/a**2)*((-4*a*(-3*a*(a + 1) + a*(3*a + 1)) + a*(-3*a*(a + 4) + a*(3*a + 4)))*(-2*a*(-2*a*(a + 1) + a*(2*a + 1)) + a*(-2*a*(a + 2) + a*(2*a + 2)))/a**2 - (-2*a*(-3*a*(a + 1) + a*(3*a + 1)) + a*(-3*a*(a + 2) + a*(3*a + 2)))*(-4*a*(-2*a*(a + 1) + a*(2*a + 1)) + a*(-2*a*(a + 4) + a*(2*a + 4)))/a**2)/a**2
```
Here, `expr.subs(a, Rational(123, 234))`  (or another rational number) returns 0 instantly. I timed `expr._random()` at 122 seconds on my laptop; it was a good workout for the CPU. (The returned value as None.)

I'm not sure what `zero` method you refer to, what module is it in? I took a look at `equals` method of Expr (which at some point is using `_random`, but only after simplification). Perhaps this is where quick option could go?  

There is also the matter of how much one is worried about erroneous classification. In case of a pivot, it's really bad to pick zero element as a pivot, but rejecting a nonzero value is no big deal; there are probably other entries that can be used instead.  For other purposes, the relative importance of two kinds of failure may be different.

asmeurer commented:
@siefkenj

siefkenj commented:
@normalhuman Did you run the test suite with `_find_reasonable_pivot` and there was no error? It'd be great if #12362 could be closed (there are several places where `_find_reasonable_pivot` should be used but currently is not).

The downside of random testing is random test failures. However, substituting values might be a faster way to determine if an expression can be guaranteed to be non-zero. Can you try adding this check to `_find_reasonable_pivot`? At the moment it looks for something provably non-zero quickly, then falls back to slower and slower methods if the previous methods fail. I'm not sure whether subbing in a number would be the fastest method of checking (beating out `.is_zero`) or whether it would be the second fastest. But I think it's worth testing (there are several bugs open about how slow `_find_reasonable_pivot` is).

Can you try hacking on `_find_reasonable_pivot` but provide a fixed list of numbers it tests instead of random ones?

@asmeurer do you have opinions on zero-checking by plugging in numbers? Is this something that should be done higher up in the sympy abstraction? (perhaps `.equals(CONSTANT)` should try subbing in a number before doing simplification? Perhaps this is already done?)

normalhuman committed:
Isolate the random-testing logic into an iszerofunc

The _find_reasonable_pivot function takes iszerofunc as a parameter.
Passing this new iszerofunc allows for much quicker, though less
reliable, determination of whether something is zero or not.
Works best for polynomials, but polynomials are the case that is
most likely to arise when computing determinants.

normalhuman commented:
@siefkenj Yes, there were no test failures if Bareiss simply uses _find_reasonable_pivot. If we are okay with the slowdown, we can just leave it at that. We could just document somewhere that three methods are available; the default one being slow but producing simple form of output, while two others are quick but messy.  (Presently, the `method` parameter of determinant is not mentioned in the docs.) 

Also, I don't think that this approach, at least in the present form, should be used more widely. It really is only suited for polynomials or rational functions that get created in the process of determinant computation. The check does not involve floating point computations: it's just a rational number being plugged in, and the check is whether the expression now is directly seen as zero. It's not really conclusive either way. But if a valid pivot gets rejected, it is usually not a big problem. 

Concerning the inclusion in  `_find_reasonable_pivot` - I noticed that this function takes iszerofunc as an argument, so I isolated all the number-plugging into a new function of that kind (called `_is_probably_zero`). This is a minimally invasive change, no rewriting of _is_reasonable_pivot needed.

The method `equals` actually tries evaluation at a random point. But it uses a floating point number, and (naturally) demands some precision from the output. This ends up being problematic because cancellation brings about loss of significance (as I remarked above in the code). In any case, this is something `equals` does only after simplification.

normalhuman committed:
Moved the generation of rational numbers to randtest.py

Added tolerance parameter to random_complex_number so it can generates
rationals with controlled size of denominator. Expanded the docstring
of _is_probably_zero to explain how it works.

normalhuman committed:
Remove unused import

smichr commented on sympy/matrices/matrices.py:
```diff
@@ -190,8 +208,11 @@ def bareiss(mat, cumm=1):
             # find a pivot and extract the remaining matrix
-            # XXX should use `_find_reasonable_pivot`.  Blocked by issue #12362
-            pivot_pos, pivot_val, _, _ = _find_pivot(mat[:, 0])
+            # Wuth the default iszerofunc,  _find_reasonable_pivot slows down
```
Wuth->With
compotation->computation

normalhuman committed:
Typos in a comment

smichr commented:
Are there any final questions or comments?

ylemkimon commented on sympy/matrices/matrices.py:
```diff
@@ -38,6 +39,31 @@ def _iszero(x):
+                    is_zero = False
+                    break
+            if is_zero is None:
+                is_zero = True
```
How about using random integers? Also I think we can add `is_constant_zero` method to the `Expr`.

normalhuman replied:
I'd be worried about something like `sin(pi*x)` if x was only tested at integers. Testing at five random rationals with 3-digit denominators gives me a lot more confidence that the expression does not just evaluate to 0 by coincidence. 

Moving it up to Expr level is a reasonable thing to consider, but I have to repeat that this isn't really a zero-testing method for general expressions. It will not detect `sin(x)**2 + cos(x)**2 - 1` as being zero, for example. It only works when x, being replaced by a rational number, gets automatically combined with surrounding rational numbers, collapsing the formula into zero. So I'm afraid it would be false advertisement to add it to Expr as a general zero tester.

I think there is a reason why this module already had a function called `is_zero`, and why `find_reasonable_pivot` provided a "zerofunc" parameter, allowing for other such functions. It's that the linear algebra manipulations have this specific need for testing many potential pivots for being zero, without necessarily obtaining a proof of them being zero.

smichr replied:
> we can add is_constant_zero method to the Expr

We have `is_constant` and `equals` already. (If we find use elsewhere then maybe a method like `rational_zero(den_digits=3, repeat=5)` which raises an error if anything other than a rational expression in the variables is being tested. For now, I agree with @normalhuman 's sentiments and would vote to leave it only here.)

smichr replied:
> How about using random integers or rationals generated from random integers

@ylemkimon , are you suggesting (as it appears to me) that using `nsimplify` in `random_complex` is overkill for what is happening? I think a more meaningful approach would be to generate the random numbers then use Rational and, optionally, make a call to the `limit_denominator` method (changing `tolerance` to `maximum_denominator`): e.g. `Random(random()).limit_denominator(1000)`. On the other hand, rather than making rational deal with the large denominators in the first place, `Rational(randint(-5000, 5000), 1000)` ... and maybe that's what you meant by your comment?

normalhuman replied:
`Rational(randint(-5000, 5000), 1000)`  would be always zero for `sin(1000*pi*x)`. Having fixed denominator is making the test less trustworthy. The original version of this PR had `Rational(randint(...) / randint(...))` but I found this approach awkward because there is a good chance of canceling the numerator and denominator. 

`nsimplify` may be expensive in general, but it shouldn't be with `rational=True`, when it's essentially a call to Python's built-in `fractions.Fraction(p, q).limit_denominator(n)`

smichr replied:
Doh...another freshman flounder. Also, I think I looked too quickly at the nsimplify function.What you have done make sense to me and it’s pretty straightforward.  Any objections to committing?

ylemkimon replied:
@smichr @normalhuman 
> We have `is_constant` and `equals` already.

> Moving it up to Expr level is a reasonable thing to consider, but I have to repeat that this isn't really a zero-testing method for general expressions.

Yes, but `equals` involves expensive `simplify` and `is_constant` just checks for whether it's a constant. I think as zero-check is used in many places, we can have special zero-check method, which implements various zero-checking methods including this one, with `method` argument or `quick` argument to control what method to use.

> are you suggesting (as it appears to me) that using nsimplify in random_complex is overkill for what is happening?

Yes, I thought just calling `randint` with fixed denominator, e.g., `10**n` or some large prime number, would suffice and be faster, but now I think, as @normalhuman suggested:

> Having fixed denominator is making the test less trustworthy.

smichr replied:
The last thing I want to think about is the possibility of random failures as was mentioned.

smichr commented:
What about generating ratios of successive primes and adding ints to expand the range if desired. This would guarantee that you are always working with a fraction. Since you are only doing 5 tests is it necessary to make the tests random? If so, you could shuffle the list before returning the Rationals.

```
>>> def r(n, dig=3):
...     p = list(primerange(10**(dig-1), 10**dig))
...     while 2*n > len(p):
...         p.append(nextprime(p[-1]))
...     return [Rational(*p[i:i+2]) for i in range(0,2*n,2)]
>>> 
>>> r(5)
[101/103,107/109,113/127,131/137,139/149]
```

normalhuman commented:
I think a deterministic check would be better. If it fails on some matrix (and any test of this kind can fail), it's better to fail every time instead of puzzling a user with a rare, non-reproducible exception. I can do [xkcd 221](https://xkcd.com/221/) and pick 5 numbers in the range [-5, 5] where the numerators and denominators are distinct primes, and denominators have 3 digits.

normalhuman committed:
Move test_at_rationals function to randtest

Isolated the testing logic to randtest. Added
randomize flag, so that testing can be made non-random.
Used non-random testing in pivot search, to make
the computation of a determinant deterministic.

smichr commented:
:-) xkcd

Have you tried doing an `expand_mul` as your simplifying function -- without doing substitution? Are there practical problem for which the substitution will outperform this? It just seems that if you are working with rational functions that the expansion should be pretty fast.

normalhuman committed:
Test by expand_mul instead of random rationals

Might be slightly slower - 88s vs 84 seconds in my test of 10 M.det()
calculations with M being
Matrix([[i + a*j + S(1)/(i+j+1) for i in range(7)] for j in range(7)])
But this is likely to have fewer incorrect results than the previous
testing method.

normalhuman commented:
I tried expand_mul. It was only slightly slower in my test (5% difference) which I think is a good price for having a more robust test. Pushed an updated with the new zerofunc. 

I didn't remove the `test_at_rationals` code from randtest, even though it's unused. Who knows, we might come up on a situation where `expand_mul`  is undesirable for some reason...

smichr commented on sympy/utilities/randtest.py:
```diff
@@ -52,6 +57,56 @@ def verify_numerically(f, g, z=None, tol=1.0e-6, a=2, b=-1, c=3, d=1):
+                return False
+    else:  # use cyclic shifts by golden ratio within [a, b]
+        counter = 0
+        golden_ratio = ((sqrt(5) + 1)/2).n(digits + 1)
```
You can spell this `GoldenRatio` (as easy as `pi`). 

BTW, anything significant about this choice?

smichr replied:
Please see #14016 -- it seems like anything that can be proved by testing with Rationals should be easy to prove with `equals` -- but I suspect there is much of which I am unaware.

normalhuman replied:
The significance of golden ratio: cyclic shift (rotation) by this account create the most uniform aperiodic pattern possible. It has to do with its continued fraction expansion. I blogged about this at https://calculus7.org/2016/01/17/irrational-sunflowers/

Re: the issue. One situation that immediately comes to mind is having a product with one very complicated tem and another which is zero. Might algebraic expansion spend a lot of time on the complicated one, before noticing the other is zero?

smichr replied:
Very nice read on the sunflowers! Thanks.

Regarding algebraic expansion and rational numbers. I suspect that rational numbers will only resolve rational expressions. For non-rational expressions [as here](https://github.com/sympy/sympy/issues/5149) substituting in rational test values results in an expression that can't be computed with precision...and we can't guess in these cases. In some cases we simply run out of precision to prove that a number is not zero. So the rationals don't help. But if the expression is rational, we should be able to expand the expression symbolically and cancel terms and return a 0 if the expression truly is zero. Will it be quicker than substituting in a rational?

```
>>> r=lambda:random_complex_number(-1,-1,1,1, rational=True)
>>> eq=(1+x)**20 - ((1+x)**20).expand()
>>> t=time();eq.expand();time()-t
0
0.188018798828125
>>> t=time();eq.subs(x,re(r()));time()-t
0
0.1770174503326416

If a number has rational real and complex parts, the result has to be separated into
real and imaginary parts to see the zero -- attempting to compute it without doing
so leads to a precisionless 0

>>> eq.subs(x,r()).n()
0.e-130 - 0.e-130*I
>>> t=time();eq.subs(x,r()).as_real_imag();time()-t
(0, 0)
0.4750478267669678
```

Can you think of a complicated expression where rational testing will work? Is it necessary to use imaginary parts when testing? I tried a higher power expression as above and the expansion out-performs substitution even more:

```
>>> eq=(1+x)**50 - ((1+x)**50).expand()
>>> t=time();eq.subs(x,re(r()));time()-t
0
0.31203126907348633
>>> t=time();eq.expand();time()-t
0
0.014001131057739258
```

BTW, that expression in #5149 could probably be made to have a radius of any size in which the result would be zero. I don't remember now where I got that expression. It seems like it must have been concocted to be that way with `x**2 + y**2 = 1`. Maybe you wouldn't run into such expressions in a single variable.

normalhuman committed:
Remove the random rational test from utilities

It was not in use anymore.
